File: ts/src/encryption.ts
// SPDX-License-Identifier: MIT
// Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci
/** Sealed transaction messages (spec/signing.md section 8): 'ZBE1' || libsodium sealed box to the recipient's converted key. */
import { blake2b } from "./crypto/blake2b.js";
import { hsalsa20, secretbox, secretboxOpen } from "./crypto/salsa.js";
import { ed25519PublicKeyToX25519, ed25519SeedToX25519, x25519, x25519Base } from "./crypto/x25519.js";
import { concat } from "./util/bytes.js";

/** The 4-byte marker in front of a sealed message field. */
export const SEALED_MAGIC = new Uint8Array([0x5a, 0x42, 0x45, 0x31]);   // "ZBE1"
/** How much longer a sealed field is than its plaintext: marker 4 + ephemeral key 32 + tag 16. */
export const SEALED_OVERHEAD = 52;

/** True when the field starts with the marker (it may still fail to open). */
export function isSealed(field: Uint8Array): boolean {
  return field.length >= 4 && field[0] === 0x5a && field[1] === 0x42 && field[2] === 0x45 && field[3] === 0x31;
}

function boxKey(sk: Uint8Array, pk: Uint8Array): Uint8Array { return hsalsa20(x25519(sk, pk), new Uint8Array(16)); }
function nonceOf(ephemeralPk: Uint8Array, recipientPkX: Uint8Array): Uint8Array { return blake2b(concat(ephemeralPk, recipientPkX), 24); }

/** Seal `plaintext` to the recipient's 32-byte Ed25519 public key. The ephemeral secret key is random unless given (tests). */
export function seal(plaintext: Uint8Array, recipientPublicKey: Uint8Array, ephemeralSecretKey?: Uint8Array): Uint8Array {
  if (recipientPublicKey.length !== 32) throw new Error("recipient public key must be 32 bytes");
  const esk = ephemeralSecretKey ?? globalThis.crypto.getRandomValues(new Uint8Array(32));
  if (esk.length !== 32) throw new Error("ephemeral secret key must be 32 bytes");
  const rpk = ed25519PublicKeyToX25519(recipientPublicKey), epk = x25519Base(esk);
  return concat(SEALED_MAGIC, epk, secretbox(boxKey(esk, rpk), nonceOf(epk, rpk), plaintext));
}

/** Open a sealed field with the recipient's 32-byte Ed25519 seed: the plaintext, or null when it is not a sealed message or the key does not open it. */
export function openSealed(field: Uint8Array, recipientSeed: Uint8Array): Uint8Array | null {
  if (!isSealed(field) || field.length < SEALED_OVERHEAD || recipientSeed.length !== 32) return null;
  const sk = ed25519SeedToX25519(recipientSeed), pk = x25519Base(sk), epk = field.subarray(4, 36);
  return secretboxOpen(boxKey(sk, epk), nonceOf(epk, pk), field.subarray(36));
}

File: rust/zbc/src/encryption.rs
// SPDX-License-Identifier: MIT
// Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci
//! Sealed transaction messages (spec/signing.md section 8): `ZBE1` || libsodium sealed box to the
//! recipient's Ed25519 key converted to X25519. X25519 from x25519-dalek, HSalsa20 from salsa20,
//! XSalsa20-Poly1305 from crypto_secretbox, the nonce from blake2.

use blake2::digest::{Update, VariableOutput};
use blake2::Blake2bVar;
use crypto_secretbox::aead::{Aead, KeyInit};
use crypto_secretbox::XSalsa20Poly1305;
use ed25519_dalek::VerifyingKey;
use rand_core::{OsRng, RngCore};
use salsa20::cipher::generic_array::GenericArray;
use sha2::{Digest, Sha512};
use x25519_dalek::{PublicKey, StaticSecret};

/// The 4-byte marker in front of a sealed message field.
pub const SEALED_MAGIC: [u8; 4] = *b"ZBE1";
/// How much longer a sealed field is than its plaintext: marker 4 + ephemeral key 32 + tag 16.
pub const SEALED_OVERHEAD: usize = 52;

/// Ed25519 public key -> X25519 public key of the same point: u = (1 + y) / (1 - y) mod p.
pub fn ed25519_public_key_to_x25519(pk: &[u8]) -> Result<[u8; 32], String> {
    let bytes: [u8; 32] = pk.try_into().map_err(|_| "public key must be 32 bytes".to_string())?;
    let vk = VerifyingKey::from_bytes(&bytes).map_err(|e| e.to_string())?;
    Ok(vk.to_montgomery().to_bytes())
}

/// Ed25519 seed -> X25519 secret key: the clamped first half of SHA-512(seed).
pub fn ed25519_seed_to_x25519(seed: &[u8]) -> [u8; 32] {
    let h = Sha512::digest(seed);
    let mut sk = [0u8; 32];
    sk.copy_from_slice(&h[..32]);
    sk[0] &= 248;
    sk[31] &= 127;
    sk[31] |= 64;
    sk
}

/// The X25519 public key of a scalar: X25519(scalar, 9).
pub fn x25519_base(scalar: &[u8; 32]) -> [u8; 32] {
    PublicKey::from(&StaticSecret::from(*scalar)).to_bytes()
}

fn box_key(sk: &[u8; 32], pk: &[u8; 32]) -> [u8; 32] {
    let shared = StaticSecret::from(*sk).diffie_hellman(&PublicKey::from(*pk));
    let key = salsa20::hsalsa::<salsa20::cipher::consts::U10>(GenericArray::from_slice(shared.as_bytes()), &GenericArray::default());
    key.into()
}

fn sealed_nonce(ephemeral_pk: &[u8; 32], recipient_pk_x: &[u8; 32]) -> [u8; 24] {
    let mut h = Blake2bVar::new(24).expect("24-byte BLAKE2b");
    h.update(ephemeral_pk);
    h.update(recipient_pk_x);
    let mut out = [0u8; 24];
    h.finalize_variable(&mut out).expect("24 bytes");
    out
}

/// True when the field starts with the marker (it may still fail to open).
pub fn is_sealed(field: &[u8]) -> bool {
    field.len() >= 4 && field[..4] == SEALED_MAGIC
}

/// Seal `plaintext` to the recipient's 32-byte Ed25519 public key. The ephemeral secret key is drawn at random unless given (tests).
pub fn seal(plaintext: &[u8], recipient_public_key: &[u8], ephemeral_secret_key: Option<&[u8; 32]>) -> Result<Vec<u8>, String> {
    let rpk = ed25519_public_key_to_x25519(recipient_public_key)?;
    let esk = match ephemeral_secret_key {
        Some(k) => *k,
        None => {
            let mut k = [0u8; 32];
            OsRng.fill_bytes(&mut k);
            k
        }
    };
    let epk = x25519_base(&esk);
    let nonce = sealed_nonce(&epk, &rpk);
    let cipher = XSalsa20Poly1305::new(GenericArray::from_slice(&box_key(&esk, &rpk)));
    let boxed = cipher.encrypt(GenericArray::from_slice(&nonce), plaintext).map_err(|_| "sealing failed".to_string())?;
    let mut out = Vec::with_capacity(SEALED_OVERHEAD + plaintext.len());
    out.extend_from_slice(&SEALED_MAGIC);
    out.extend_from_slice(&epk);
    out.extend_from_slice(&boxed);
    Ok(out)
}

/// Open a sealed field with the recipient's 32-byte Ed25519 seed: the plaintext, or None when it is not a sealed message or the key does not open it.
pub fn open_sealed(field: &[u8], recipient_seed: &[u8]) -> Option<Vec<u8>> {
    if !is_sealed(field) || field.len() < SEALED_OVERHEAD || recipient_seed.len() != 32 {
        return None;
    }
    let sk = ed25519_seed_to_x25519(recipient_seed);
    let pk = x25519_base(&sk);
    let epk: [u8; 32] = field[4..36].try_into().ok()?;
    let cipher = XSalsa20Poly1305::new(GenericArray::from_slice(&box_key(&sk, &epk)));
    cipher.decrypt(GenericArray::from_slice(&sealed_nonce(&epk, &pk)), &field[36..]).ok()
}
